### PR TITLE
Map module and communication

### PR DESCRIPTION
**Pull request by Henkecool**
# Map module and communication

* Add Map, Node, Connection, DIrection and Position datatypes
* Add Map Route-module
* Add Map View-feature
* Add corresponding controllers
* Extend Server to be able to handle late invoking of methods concerning respons listening.
* Extend Car, ResquestBuilder and Server with relevant methods for server communication.

## Relevant issues
#3 WIP

## Testing
Integration tested with central-module.

### CODE REVIEW

File: src/gui/controllers/MapRouteModuleController.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     * Handle JavaFX Map Node-button click. This gets the corresponding Map Node and adds it to the current route if
+     * currently buildning route and if the clicked node is valid (not the same as last node).
+     * @param mouseEvent Triggering event.
+     */
+    private void handleNodeClick(MouseEvent mouseEvent) {
+        if (buildingRoute.get()) {
+            mouseEvent.consume();
+
+            Circle fxNode = (Circle) mouseEvent.getSource();
+            Node node = Car.getInstance().map.get().getNode(Integer.parseInt(String.valueOf(fxNode.getId().charAt(10))));
```
Comment (angloth) [inline src/gui/controllers/MapRouteModuleController.java]: Java LOL

File: src/gui/controllers/MapViewFeatureController.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     */
+    public void update(Map data) {
+        redraw(data);
+    }
+
+    /**
+     * Draw the given Map to the view.
+     * @param map Map data
+     */
+    private void redraw(Map map) {
```
Comment (angloth) [inline src/gui/controllers/MapViewFeatureController.java:47]: Doesn't clear the current map, maybe it should just be called draw or there is missed implementation?

File: src/map/Map.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    public int byteSize() {
+        int sum = 2;
+        for (Node n : nodes) {
+            sum += n.byteSize();
+        }
+        return sum;
+    }
+
+    public byte[] toBytes() {
```
Comment (angloth) [inline src/map/Map.java:43]: The methods byteSize() and toBytes() are shared in many classes and seem to fit being from an Interface, not necessary though.

**Review by angloth — APPROVED** [review]
Looks good, just left some questions.

File: src/gui/controllers/MapRouteModuleController.java (hunk 1 above)
Comment (Henkecool) [reply src/gui/controllers/MapRouteModuleController.java]: Fixed! (tho still Java)

File: src/gui/controllers/MapViewFeatureController.java (hunk 2 above)
Comment (Henkecool) [reply src/gui/controllers/MapViewFeatureController.java:47]: Missed clearing of layers, fixed now

File: src/map/Map.java (hunk 3 above)
Comment (Henkecool) [reply src/map/Map.java:43]: yeeeehh no (but yeah)

**Pull request merged**